cli/app.py

```diff
@@ -0,0 +1,155 @@
+#    Nordic Semiconductor ASA integrated circuit.
+#
+# 5. Any software provided in binary form under this license must not be reverse
+#    engineered, decompiled, modified and/or disassembled.
+#
+# THIS SOFTWARE IS PROVIDED BY NORDIC SEMICONDUCTOR ASA "AS IS" AND ANY EXPRESS
+# OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES
+# OF MERCHANTABILITY, NONINFRINGEMENT, AND FITNESS FOR A PARTICULAR PURPOSE ARE
+# DISCLAIMED. IN NO EVENT SHALL NORDIC SEMICONDUCTOR ASA OR CONTRIBUTORS BE
+# LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+# CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE
+# GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
+# HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+# LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT
+# OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+import argparse
+import time
+from SnifferAPI import Sniffer, UART
+
+def setup(capture_file):
+    ports = UART.find_sniffer()
+
+    if len(ports) > 0:
+        sniffer = Sniffer.Sniffer(portnum=ports[0], baudrate=1000000, capture_file_path=capture_file)
+        sniffer.setSupportedProtocolVersion(2)
+
+    else:
+        print("No sniffers found!")
+        return
+
+    sniffer.start()
+    return sniffer
+
+
+def scan(sniffer, timeout, address=None, name=None):
+    sniffer.scan()
+    devices = None
+    for _ in range(timeout):
+        time.sleep(1)
+        devices = sniffer.getDevices()
+        if address is not None:
+            device = _find_device_by_address(devices, address)
+            if device is not None:
+                return device
+
+        elif name is not None:
+            device = devices.find(name)
+            if device is not None:
+                return device
+
+    return devices
+
+
+def _find_device_by_address(devices, address):
+    for device in devices.asList():
+        if _address_to_string(device) == address.replace(':', ''):
+            return device
+
+
+def follow(sniffer, device):
+    sniffer.follow(device)
+    loop(sniffer)
+
+
+def loop(sniffer):
+    # Enter main loop
+    loops = 0
+    packet_count = 0
+    while True:
+        time.sleep(0.1)
+        # Get (pop) unprocessed BLE packets.
+        packets = sniffer.getPackets()
+        packet_count = packet_count + len(packets)
+        loops += 1
+
+        # print diagnostics every so often
+        if loops % 20 == 0:
+            print("inConnection", sniffer.inConnection)
+            print("currentConnectRequest", sniffer.currentConnectRequest)
+            print("packetsInLastConnection", sniffer.packetsInLastConnection)
+            print("packetCount", packet_count)
+            print()
+
+
+def main():
+    parser = argparse.ArgumentParser(description='Sniff Bluetooth LE traffic over the air.')
+    parser.add_argument('--scan', '-s',
```

Comment:
Seems that there should be two sub commands, scan and sniff. `--address`, `--name` and `--capture-file` would belong as parameters to sniff command.

Reply:
Makes sense. I will add `--sniff` argument too. The `--capture-file` argument actually has affect with `--scan` as the pcap is saved during scanning too.

Reply:
I had sub parsers in mind, https://docs.python.org/3/library/argparse.html#argparse.ArgumentParser.add_subparsers.

The API would then be something like:
`nrf-sniffer-cli --capture-file sniff --address/--name`
`nrf-sniffer-cli --capture-file scan`

Should be no need for the extra if conditions with sub parsers.

Reply:
Ah, I wasn't aware of sub parsers, seems useful.